elasticsearch/store/store.py

```diff
@@ -0,0 +1,399 @@
+        custom_query: Optional[Callable[[Dict, Optional[str]], Dict]] = None,
+    ) -> List[Dict[str, Any]]:
+        """
+        Args:
+            query: Input query string.
+            query_vector: Input embedding vector. If given, input query string is
+                ignored.
+            k: Number of returned results.
+            num_candidates: Number of candidates to fetch from data nodes in knn.
+            fields: List of field names to return.
+            filter: Elasticsearch filters to apply.
+            custom_query: Function to modify the Elasticsearch query body before it is
+                sent to Elasticsearch.
+
+        Returns:
+            List of document hits. Includes _index, _id, _score and _source.
+        """
+        if fields is None:
+            fields = []
+        if "metadata" not in fields:
+            fields.append("metadata")
+        if self.text_field not in fields:
+            fields.append(self.text_field)
+
+        query_body = self.retrieval_strategy.es_query(
+            query=query,
+            k=k,
+            num_candidates=num_candidates,
+            filter=filter or [],
+            query_vector=query_vector,
+        )
+        logger.debug(f"Query body: {query_body}")
+
+        if custom_query is not None:
+            query_body = custom_query(query_body, query)
+            logger.debug(f"Calling custom_query, Query body now: {query_body}")
+
+        response = await self.es_client.search(
+            index=self.index_name,
+            **query_body,
+            size=k,
+            source=True,
+            source_includes=fields,
+        )
+
+        return response["hits"]["hits"]
+
+    def search(
+        self,
+        query: Optional[str],
+        query_vector: Optional[List[float]] = None,
+        k: int = 4,
+        num_candidates: int = 50,
+        fields: Optional[List[str]] = None,
+        filter: Optional[List[dict]] = None,
+        custom_query: Optional[Callable[[Dict, Optional[str]], Dict]] = None,
+    ) -> List[Dict[str, Any]]:
+        return asyncio.get_event_loop().run_until_complete(
+            self.search_async(
+                query=query,
+                query_vector=query_vector,
+                k=k,
+                num_candidates=num_candidates,
+                fields=fields,
+                filter=filter,
+                custom_query=custom_query,
+            )
+        )
+
+    async def _create_index_if_not_exists(self) -> None:
+        exists = await self.es_client.indices.exists(index=self.index_name)
+        if exists.meta.status == 200:
+            logger.debug(f"Index {self.index_name} already exists. Skipping creation.")
+        else:
+            await self.retrieval_strategy.create_index(
+                client=self.es_client,
+                index_name=self.index_name,
+                metadata_mapping=self.metadata_mapping,
+            )
+
+    def max_marginal_relevance_search(
```

Comment:
OK, I guess I was wondering if this was meant to run async at all since you're only calling `embed_query` and not `embed_query_async` here. If we do want to run this async, we may have to use tricks depending on how long this typically takes.

Reply:
The computationally expensive part in Python can be limited by limiting `k` and `num_candidates`. We could think of sensible maximum values to enforce here.

Reply:
For reference, slow in the context of asyncio depends on the application but a good baseline is 100ms: https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.slow_callback_duration. If it usually takes more time than that, we can think about mitigation measures.